File: TCNs_projection_based.m
%___________________________________________________________________________________________________________________________________________________%
%
% Twisted Convolutional Networks (TCNs) source codes (version 1.1)
% Projection-based TCN Model
% Website and codes of TCN: https://github.com/junbolian/Twisted_Convolutional_Networks
% Last update: Sep 29 2025
% E-Mail: jacoblian@u.northwestern.edu
% Author: Junbo Jacob Lian 
%____________________________________________________________________________________________________________________________________________________%

%% Twisted Convolutional Networks (TCNs) for Dataset with Arbitrary Features

%% Clean Environment
clear; close all; clc;

%% Set Random Seed for Reproducibility
rng(42);

%% Load Dataset
filename = 'dataset.xlsx';
data = readmatrix(filename);

%% Shuffle Features and Separate Labels
labels   = data(:, end);
features = data(:, 1:end-1);
[num_samples, num_features] = size(features);

% Record original feature indices for naming (for explainability)
orig_feature_idx = 1:num_features;

% Shuffle the feature columns to increase independence between adjacent features
shuffle_idx = randperm(num_features);
features = features(:, shuffle_idx);
orig_feature_idx = orig_feature_idx(shuffle_idx);  % keep mapping for names

%% Shuffle Data Rows
idx = randperm(size(features, 1));
features = features(idx, :);
labels   = labels(idx);

%% Feature Combination Configuration
num_combinations = 2; % C (>=2). Here we use pairs by default.

%% Feature Combination Method Selection
combination_method = 'pairwise'; % 'multiplicative' or 'pairwise'

%% Create Combinations of Features (choose C features each)
combinations = nchoosek(1:num_features, num_combinations);    % [num_combined_features, C]
num_combined_features = size(combinations, 1);

%% Generate Combined Feature Data  (this is z = concat{phi_S(x)})
combined_data = zeros(num_samples, num_combined_features);
switch lower(combination_method)
    case 'multiplicative' % Approach I: product of selected features
        for i = 1:num_combined_features
            idxs = combinations(i, :);
            combined_data(:, i) = prod(features(:, idxs), 2);
        end
    case 'pairwise'       % Approach II: sum of pairwise products within the subset
        for i = 1:num_combined_features
            idxs = combinations(i, :);
            s = zeros(num_samples,1);
            for j = 1:length(idxs)
                for k = (j+1):length(idxs)
                    s = s + features(:, idxs(j)) .* features(:, idxs(k));
                end
            end
            combined_data(:, i) = s;
        end
    otherwise
        error('Unknown combination_method.');
end

%% Split Data into Training and Testing Sets
cv = cvpartition(labels, 'HoldOut', 0.3);
train_data   = combined_data(training(cv), :);
test_data    = combined_data(test(cv),     :);
train_labels = labels(training(cv));
test_labels  = labels(test(cv));

%% Convert Labels to Categorical
train_labels = categorical(train_labels);
test_labels  = categorical(test_labels);

%% Determine Number of Classes
num_classes = numel(categories(categorical(labels)));

%% Hidden sizes (configurable)
H1 = 64;   % first hidden width
H2 = 256;  % second hidden / add-output width (can be smaller than num_combined_features)

%% ===== TCN Model Definition with PROJECTION-BASED residual =====
% Main branch (before addition)
mainLayers = [
    featureInputLayer(num_combined_features, 'Normalization', 'zscore', 'Name', 'input')
    
    fullyConnectedLayer(H1, 'WeightsInitializer', 'he', 'Name', 'fc1')
    batchNormalizationLayer('Name', 'bn1')
    reluLayer('Name', 'relu1')
    
    fullyConnectedLayer(H2, 'WeightsInitializer', 'he', 'Name', 'fc2')
    batchNormalizationLayer('Name', 'bn2')
    reluLayer('Name', 'relu2')
];

% Post-addition layers (from addition layer onwards)
postAddLayers = [
    additionLayer(2, 'Name', 'add')      % add/in1 = main; add/in2 = projected skip
    reluLayer('Name', 'post_add_relu')
    
    % Classification head
    fullyConnectedLayer(10, 'WeightsInitializer', 'he', 'Name', 'fc_head')
    dropoutLayer(0.5, 'Name', 'dropout1')
    fullyConnectedLayer(num_classes, 'Name', 'output_fc')
    softmaxLayer('Name', 'softmax')
    classificationLayer('Name', 'class_output')
];

% Build graph
lgraph = layerGraph(mainLayers);
lgraph = addLayers(lgraph, postAddLayers);

% Residual source: choose 'input' (from combination layer output z) or 'relu1'
residual_source = 'input';  % 'input' is closer to "block input"; 'relu1' is also reasonable.

% Projection layers for residual connection (always include; if dims match it behaves near-identity)
projLayers = [
    fullyConnectedLayer(H2, 'WeightsInitializer', 'he', ...
        'BiasLearnRateFactor', 0, 'Name', 'proj_fc')
    batchNormalizationLayer('Name', 'proj_bn') % optional; remove if you prefer no BN
];
lgraph = addLayers(lgraph, projLayers);

% Connect paths
lgraph = connectLayers(lgraph, 'relu2', 'add/in1');      % main -> add
lgraph = connectLayers(lgraph, residual_source, 'proj_fc');
lgraph = connectLayers(lgraph, 'proj_bn', 'add/in2');    % proj skip -> add

%% Training Options
options = trainingOptions('adam', ...
    'InitialLearnRate', 1e-3, ...
    'MaxEpochs', 200, ...
    'MiniBatchSize', 32, ...
    'Shuffle', 'every-epoch', ...
    'ValidationData', {test_data, test_labels}, ...
    'ValidationFrequency', 30, ...
    'Verbose', false, ...
    'Plots', 'training-progress', ...
    'L2Regularization', 1e-4);

%% Train TCN Model  (return training info for custom curves)
[trained_tcn, info] = trainNetwork(train_data, train_labels, lgraph, options);

%% ===== Pretty Training Curves: Loss & Accuracy (Train vs Val) =====
% Some MATLAB releases name fields slightly differently; guard with isfield.
epochs = 1:numel(info.TrainingLoss);
% Smooth curves for nicer visuals
S = 5;  % smoothing window
trLoss = smoothdata(info.TrainingLoss, 'movmean', S);
vaLoss = smoothdata(info.ValidationLoss, 'movmean', S);
trAcc  = smoothdata(info.TrainingAccuracy, 'movmean', S);
vaAcc  = smoothdata(info.ValidationAccuracy, 'movmean', S);

%% ===== Training Curves in ONE figure (Acc top, Loss bottom) =====
% Build epoch axis & pad with NaN to equal length
N = numel(info.TrainingLoss);
pad = @(v) [v(:).' nan(1, max(0, N-numel(v)))];

trLoss = pad(info.TrainingLoss);
vaLoss = pad(info.ValidationLoss);
trAcc  = pad(info.TrainingAccuracy);
vaAcc  = pad(info.ValidationAccuracy);

% Smooth (moving mean) while keeping NaNs
W = 7;  % smoothing window
trLossS = smoothdata(trLoss, 'movmean', W, 'omitnan');
vaLossS = smoothdata(vaLoss, 'movmean', W, 'omitnan');
trAccS  = smoothdata(trAcc , 'movmean', W, 'omitnan');
vaAccS  = smoothdata(vaAcc , 'movmean', W, 'omitnan');

epochs = 1:N;

% Okabe–Ito colorblind-safe palette
cAccTr = [0 114 178]/255;  % blue
cAccVa = [213 94 0]/255;   % vermillion
cLosTr = [0 158 115]/255;  % bluish green
cLosVa = [204 121 167]/255;% reddish purple

lw   = 2.2;     % line width
ms   = 5.5;     % marker size
mkEvery = max(2, round(N/12));              
mkIdx   = 1:mkEvery:N;                       

% Figure aesthetics (Nature-like)
fig = figure('Color','w','Units','pixels','Position',[100 100 760 760]);
tiledlayout(2,1,'Padding','compact','TileSpacing','compact');

% Helper to unify axes style
setAxesNature = @(ax) set(ax, ...
    'FontName','Helvetica', 'FontSize',11, ...
    'LineWidth',1, 'TickDir','out', ...
    'Box','off', 'XMinorTick','on', 'YMinorTick','on');

% ---- Accuracy (top) ----
ax1 = nexttile;
p1 = plot(epochs, trAccS, '-', 'Color', cAccTr, 'LineWidth', lw); hold on;
p2 = plot(epochs, vaAccS, '--', 'Color', cAccVa, 'LineWidth', lw, ...
          'Marker','o','MarkerSize',ms,'MarkerIndices',mkIdx, ...
          'MarkerFaceColor','w','MarkerEdgeColor',cAccVa);
ylabel('Accuracy (%)');
title('Training/Validation Accuracy (top) and Loss (bottom)');
grid(ax1,'on'); setAxesNature(ax1);


leg1 = legend([p1 p2], {'Train Acc','Val Acc'}, 'Location','northeast');
leg1.Box = 'off';

% ---- Loss (bottom) ----
ax2 = nexttile;
p3 = plot(epochs, trLossS, '-',  'Color', cLosTr, 'LineWidth', lw); hold on;
p4 = plot(epochs, vaLossS, '--', 'Color', cLosVa, 'LineWidth', lw, ...
          'Marker','s','MarkerSize',ms,'MarkerIndices',mkIdx, ...
          'MarkerFaceColor','w','MarkerEdgeColor',cLosVa);
xlabel('Epoch'); ylabel('Loss');
grid(ax2,'on'); setAxesNature(ax2);

leg2 = legend([p3 p4], {'Train Loss','Val Loss'}, 'Location','northeast');
leg2.Box = 'off';

xlim(ax1, [1 N]); xlim(ax2, [1 N]);

% ===== High-quality export (choose one) =====
% exportgraphics(fig, 'train_curves.pdf',  'ContentType','vector');  
% exportgraphics(fig, 'train_curves.tif',  'Resolution', 600);    
% exportgraphics(fig, 'train_curves.png',  'Resolution', 300);


%% Evaluate TCN Model
predicted_labels = classify(trained_tcn, test_data);
accuracy = mean(predicted_labels == test_labels);
fprintf('Test Accuracy: %.2f%%\n', accuracy * 100);

%% Enhanced Evaluation Metrics
conf_matrix = confusionmat(test_labels, predicted_labels);
figure('Color','w');
confusionchart(test_labels, predicted_labels);
title('Confusion Matrix');

precision = diag(conf_matrix) ./ sum(conf_matrix, 2);
recall    = diag(conf_matrix) ./ sum(conf_matrix, 1)';
precision(isnan(precision)) = 0;
recall(isnan(recall))       = 0;

f1_score = 2 * (precision .* recall) ./ (precision + recall);
f1_score(isnan(f1_score)) = 0;

for i = 1:num_classes
    fprintf('Class %d - Precision: %.2f, Recall: %.2f, F1-score: %.2f\n', ...
        i, precision(i) * 100, recall(i) * 100, f1_score(i) * 100);
end

avg_precision = mean(precision);
avg_recall    = mean(recall);
avg_f1        = mean(f1_score);
fprintf('Average Precision: %.2f%%\n', avg_precision * 100);
fprintf('Average Recall: %.2f%%\n',    avg_recall    * 100);
fprintf('Average F1-score: %.2f%%\n',  avg_f1        * 100);

%% ROC Curve and AUC
scores = predict(trained_tcn, test_data);
figure('Color','w'); hold on;
present = false(1, num_classes);
for i = 1:num_classes
    pres = any(test_labels == categorical(i));
    present(i) = pres;
    if pres
        [X, Y, T, AUC] = perfcurve(test_labels == categorical(i), scores(:, i), true);
        plot(X, Y, 'LineWidth', 1.8);
        fprintf('Class %d - AUC: %.2f\n', i, AUC);
    else
        fprintf('Class %d - Not present in test data, skipping AUC calculation\n', i);
    end
end
plot([0 1],[0 1],':','LineWidth',1); % diagonal reference
title('ROC Curves');
xlabel('False Positive Rate'); ylabel('True Positive Rate');
legend(arrayfun(@(x) sprintf('Class %d', x), find(present), 'UniformOutput', false), ...
    'Location','southeastoutside');
grid on; box on; hold off;

%% ===== Explainability: Input × Gradient Attribution (no new figures) =====
% Goal: quantify each explicit interaction feature's contribution to the chosen class.

% 1) Choose class for attribution (dominant class on test set)
[~, pred_cls_idx] = max(mean(scores, 1));  % 1..num_classes
class_idx = pred_cls_idx;  % or set manually, e.g., class_idx = 1;

% 2) Compute finite-difference gradients wrt input features (test set)
num_test_samples   = size(test_data, 1);
num_combined_feats = size(test_data, 2);
gradients = zeros(num_test_samples, num_combined_feats, 'like', test_data);
epsilon   = 1e-4;

for i = 1:num_test_samples
    base_input  = test_data(i, :);
    base_scores = predict(trained_tcn, base_input);
    base_score  = base_scores(class_idx);
    for j = 1:num_combined_feats
        perturbed = base_input;
        perturbed(j) = perturbed(j) + epsilon;
        pert_scores = predict(trained_tcn, perturbed);
        gradients(i, j) = (pert_scores(class_idx) - base_score) / epsilon;
    end
end

% 3) Input×Grad attribution and normalization
input_grad_product = test_data .* gradients;
E_abs    = mean(abs(input_grad_product), 1);
E_signed = mean(input_grad_product, 1);
den = sum(E_abs) + eps;
I = E_abs / den;   % normalized attribution share per interaction feature

% 4) Human-readable names for combinations (using original indices)
comb_names = cell(num_combined_features,1);
for i = 1:num_combined_features
    idxs = combinations(i,:);          % indices in shuffled feature space
    orig_idx = orig_feature_idx(idxs); % back to original feature indices
    if num_combinations == 2
        comb_names{i} = sprintf('(%d,%d)', orig_idx(1), orig_idx(2));
    else
        comb_names{i} = sprintf('(%s)', strjoin(string(orig_idx), ','));
    end
end

% 5) Top-K list and coverage
K = min(10, num_combined_features);      % Top-10 by default
[sorted_I, order] = sort(I, 'descend');
top_idx = order(1:K);
top_cov = 100 * sum(sorted_I(1:K));

fprintf('\n==== Explainability (class %d) - Input×Grad Attribution ====\n', class_idx);
fprintf('Top-%d coverage of attribution: %.1f%% of total\n', K, top_cov);
for t = 1:K
    j = top_idx(t);
    fprintf('#%d  S=%s   importance=%.1f%%   signed_contribution=%.4f\n', ...
        t, comb_names{j}, 100*I(j), E_signed(j));
end

File: TCNs_projection_based.py
"""
Twisted Convolutional Networks (TCNs): Enhancing Feature Interactions for Non-Spatial Data Classification
Projection-based TCN Model
Original MATLAB code converted to Python using TensorFlow/Keras
Author: Junbo Jacob Lian
Date: Sep 29, 2025
https://github.com/junbolian/Twisted-Convolutional-Networks
"""

import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.model_selection import train_test_split
from sklearn.metrics import confusion_matrix, precision_recall_fscore_support, roc_curve, auc
from sklearn.preprocessing import LabelBinarizer, StandardScaler
from itertools import combinations as comb
import tensorflow as tf
from tensorflow import keras
from tensorflow.keras import layers, models, optimizers, callbacks
from tensorflow.keras.utils import to_categorical
import warnings
warnings.filterwarnings('ignore')

# Set random seed for reproducibility
np.random.seed(42)
tf.random.set_seed(42)

class TCNClassifier:
    def __init__(self, num_combinations=2, combination_method='pairwise', H1=64, H2=256):
        """
        Initialize TCN Classifier
        
        Parameters:
        num_combinations: int, number of features to combine (C >= 2)
        combination_method: str, 'multiplicative' or 'pairwise'
        H1: int, first hidden layer width
        H2: int, second hidden layer/add-output width
        """
        self.num_combinations = num_combinations
        self.combination_method = combination_method
        self.H1 = H1
        self.H2 = H2
        self.model = None
        self.scaler = StandardScaler()
        self.feature_combinations = None
        self.orig_feature_idx = None
        
    def load_and_prepare_data(self, filename):
        """Load dataset from Excel file and prepare features/labels"""
        # Load data
        data = pd.read_excel(filename, header=None).values
        
        # Separate features and labels
        self.labels = data[:, -1].astype(int)
        features = data[:, :-1]
        num_samples, num_features = features.shape
        
        # Record original feature indices for explainability
        self.orig_feature_idx = np.arange(num_features)
        
        # Shuffle feature columns to increase independence
        shuffle_idx = np.random.permutation(num_features)
        features = features[:, shuffle_idx]
        self.orig_feature_idx = self.orig_feature_idx[shuffle_idx]
        
        # Shuffle data rows
        idx = np.random.permutation(num_samples)
        self.features = features[idx]
        self.labels = self.labels[idx]
        
        return self.features, self.labels
    
    def create_feature_combinations(self):
        """Create combinations of features"""
        num_features = self.features.shape[1]
        self.feature_combinations = list(comb(range(num_features), self.num_combinations))
        num_combined_features = len(self.feature_combinations)
        
        # Generate combined feature data
        num_samples = self.features.shape[0]
        combined_data = np.zeros((num_samples, num_combined_features))
        
        if self.combination_method.lower() == 'multiplicative':
            # Approach I: product of selected features
            for i, idxs in enumerate(self.feature_combinations):
                combined_data[:, i] = np.prod(self.features[:, idxs], axis=1)
        elif self.combination_method.lower() == 'pairwise':
            # Approach II: sum of pairwise products within the subset
            for i, idxs in enumerate(self.feature_combinations):
                s = np.zeros(num_samples)
                for j in range(len(idxs)):
                    for k in range(j + 1, len(idxs)):
                        s += self.features[:, idxs[j]] * self.features[:, idxs[k]]
                combined_data[:, i] = s
        else:
            raise ValueError(f"Unknown combination_method: {self.combination_method}")
        
        return combined_data
    
    def build_tcn_model(self, input_dim, num_classes):
        """Build TCN model with projection-based residual connection"""
        # Input layer
        input_layer = layers.Input(shape=(input_dim,), name='input')
        
        # Normalize input
        normalized = layers.BatchNormalization(name='input_norm')(input_layer)
        
        # Main branch
        x = layers.Dense(self.H1, kernel_initializer='he_normal', name='fc1')(normalized)
        x = layers.BatchNormalization(name='bn1')(x)
        x = layers.ReLU(name='relu1')(x)
        
        x = layers.Dense(self.H2, kernel_initializer='he_normal', name='fc2')(x)
        x = layers.BatchNormalization(name='bn2')(x)
        x = layers.ReLU(name='relu2')(x)
        
        # Projection branch for residual connection
        # Project input to match H2 dimensions
        proj = layers.Dense(self.H2, kernel_initializer='he_normal', 
                           use_bias=False, name='proj_fc')(normalized)
        proj = layers.BatchNormalization(name='proj_bn')(proj)
        
        # Addition layer (residual connection)
        added = layers.Add(name='add')([x, proj])
        added = layers.ReLU(name='post_add_relu')(added)
        
        # Classification head
        x = layers.Dense(10, kernel_initializer='he_normal', name='fc_head')(added)
        x = layers.Dropout(0.5, name='dropout1')(x)
        x = layers.Dense(num_classes, name='output_fc')(x)
        output = layers.Softmax(name='softmax')(x)
        
        # Create model
        model = models.Model(inputs=input_layer, outputs=output, name='TCN')
        
        return model
    
    def train(self, X_train, y_train, X_val, y_val, epochs=200, batch_size=32, lr=1e-3):
        """Train the TCN model"""
        # Get number of classes
        num_classes = len(np.unique(np.concatenate([y_train, y_val])))
        
        # Convert labels to categorical
        y_train_cat = to_categorical(y_train, num_classes)
        y_val_cat = to_categorical(y_val, num_classes)
        
        # Normalize features
        X_train = self.scaler.fit_transform(X_train)
        X_val = self.scaler.transform(X_val)
        
        # Build model
        self.model = self.build_tcn_model(X_train.shape[1], num_classes)
        
        # Compile model
        optimizer = optimizers.Adam(learning_rate=lr)
        self.model.compile(
            optimizer=optimizer,
            loss='categorical_crossentropy',
            metrics=['accuracy']
        )
        
        # Callbacks
        early_stop = callbacks.EarlyStopping(
            monitor='val_loss', 
            patience=30, 
            restore_best_weights=True
        )
        
        reduce_lr = callbacks.ReduceLROnPlateau(
            monitor='val_loss',
            factor=0.5,
            patience=15,
            min_lr=1e-7
        )
        
        # Train model
        history = self.model.fit(
            X_train, y_train_cat,
            validation_data=(X_val, y_val_cat),
            epochs=epochs,
            batch_size=batch_size,
            callbacks=[early_stop, reduce_lr],
            verbose=1
        )
        
        return history
    
    def evaluate(self, X_test, y_test):
        """Evaluate model performance"""
        # Normalize test data
        X_test_norm = self.scaler.transform(X_test)
        
        # Get predictions
        y_pred_proba = self.model.predict(X_test_norm)
        y_pred = np.argmax(y_pred_proba, axis=1)
        
        # Calculate accuracy
        accuracy = np.mean(y_pred == y_test)
        print(f'Test Accuracy: {accuracy * 100:.2f}%')
        
        # Calculate precision, recall, F1
        precision, recall, f1, _ = precision_recall_fscore_support(
            y_test, y_pred, average=None, zero_division=0
        )
        
        # Print metrics for each class
        num_classes = len(np.unique(y_test))
        for i in range(num_classes):
            print(f'Class {i} - Precision: {precision[i]*100:.2f}, '
                  f'Recall: {recall[i]*100:.2f}, F1-score: {f1[i]*100:.2f}')
        
        # Average metrics
        print(f'Average Precision: {np.mean(precision)*100:.2f}%')
        print(f'Average Recall: {np.mean(recall)*100:.2f}%')
        print(f'Average F1-score: {np.mean(f1)*100:.2f}%')
        
        return y_pred, y_pred_proba
    
    def plot_training_curves(self, history):
        """Plot training and validation curves"""
        # Smooth function
        def smooth_curve(points, factor=0.7):
            smoothed = []
            for point in points:
                if smoothed:
                    previous = smoothed[-1]
                    smoothed.append(previous * factor + point * (1 - factor))
                else:
                    smoothed.append(point)
            return smoothed
        
        # Get data
        epochs = range(1, len(history.history['loss']) + 1)
        train_loss = smooth_curve(history.history['loss'])
        val_loss = smooth_curve(history.history['val_loss'])
        train_acc = smooth_curve(history.history['accuracy'])
        val_acc = smooth_curve(history.history['val_accuracy'])
        
        # Create figure
        fig, (ax1, ax2) = plt.subplots(2, 1, figsize=(10, 10))
        
        # Accuracy subplot
        ax1.plot(epochs, train_acc, 'b-', label='Train Accuracy', linewidth=2)
        ax1.plot(epochs, val_acc, 'r--', label='Validation Accuracy', 
                linewidth=2, marker='o', markevery=10, markersize=5)
        ax1.set_ylabel('Accuracy')
        ax1.set_title('Training/Validation Accuracy and Loss')
        ax1.legend(loc='lower right')
        ax1.grid(True, alpha=0.3)
        
        # Loss subplot
        ax2.plot(epochs, train_loss, 'g-', label='Train Loss', linewidth=2)
        ax2.plot(epochs, val_loss, 'm--', label='Validation Loss', 
                linewidth=2, marker='s', markevery=10, markersize=5)
        ax2.set_xlabel('Epoch')
        ax2.set_ylabel('Loss')
        ax2.legend(loc='upper right')
        ax2.grid(True, alpha=0.3)
        
        plt.tight_layout()
        plt.show()
        
    def plot_confusion_matrix(self, y_true, y_pred):
        """Plot confusion matrix"""
        cm = confusion_matrix(y_true, y_pred)
        
        plt.figure(figsize=(8, 6))
        sns.heatmap(cm, annot=True, fmt='d', cmap='Blues', square=True)
        plt.ylabel('True Label')
        plt.xlabel('Predicted Label')
        plt.title('Confusion Matrix')
        plt.show()
        
    def plot_roc_curves(self, y_test, y_pred_proba):
        """Plot ROC curves for each class"""
        # Binarize labels for multi-class ROC
        lb = LabelBinarizer()
        y_test_bin = lb.fit_transform(y_test)
        n_classes = y_test_bin.shape[1]
        
        plt.figure(figsize=(10, 8))
        
        for i in range(n_classes):
            if np.any(y_test_bin[:, i]):
                fpr, tpr, _ = roc_curve(y_test_bin[:, i], y_pred_proba[:, i])
                roc_auc = auc(fpr, tpr)
                plt.plot(fpr, tpr, linewidth=2, 
                        label=f'Class {i} (AUC = {roc_auc:.2f})')
                print(f'Class {i} - AUC: {roc_auc:.2f}')
            else:
                print(f'Class {i} - Not present in test data, skipping AUC calculation')
        
        plt.plot([0, 1], [0, 1], 'k--', linewidth=1)
        plt.xlabel('False Positive Rate')
        plt.ylabel('True Positive Rate')
        plt.title('ROC Curves')
        plt.legend(loc='lower right')
        plt.grid(True, alpha=0.3)
        plt.show()
        
    def explain_features(self, X_test, y_test, class_idx=None):
        """
        Feature importance using gradient-based attribution
        """
        X_test_norm = self.scaler.transform(X_test)
        
        if class_idx is None:
            # Use dominant class
            y_pred_proba = self.model.predict(X_test_norm)
            class_idx = np.argmax(np.mean(y_pred_proba, axis=0))
        
        # Convert to TensorFlow tensors
        X_tensor = tf.convert_to_tensor(X_test_norm, dtype=tf.float32)
        
        # Compute gradients
        with tf.GradientTape() as tape:
            tape.watch(X_tensor)
            predictions = self.model(X_tensor, training=False)
            class_output = predictions[:, class_idx]
        
        gradients = tape.gradient(class_output, X_tensor)
        
        # Input × Gradient attribution
        input_grad_product = X_test_norm * gradients.numpy()
        E_abs = np.mean(np.abs(input_grad_product), axis=0)
        E_signed = np.mean(input_grad_product, axis=0)
        
        # Normalize
        I = E_abs / (np.sum(E_abs) + 1e-10)
        
        # Create feature names
        comb_names = []
        for i, idxs in enumerate(self.feature_combinations):
            orig_idx = self.orig_feature_idx[list(idxs)]
            if self.num_combinations == 2:
                comb_names.append(f'({orig_idx[0]},{orig_idx[1]})')
            else:
                comb_names.append(f'({",".join(map(str, orig_idx))})')
        
        # Top-K features
        K = min(10, len(comb_names))
        top_idx = np.argsort(I)[::-1][:K]
        top_cov = 100 * np.sum(I[top_idx])
        
        print(f'\n==== Explainability (class {class_idx}) - Input×Grad Attribution ====')
        print(f'Top-{K} coverage of attribution: {top_cov:.1f}% of total')
        
        for t, j in enumerate(top_idx):
            print(f'#{t+1}  S={comb_names[j]}   importance={100*I[j]:.1f}%   '
                  f'signed_contribution={E_signed[j]:.4f}')


def main():
    """Main execution function"""
    # Initialize TCN
    tcn = TCNClassifier(
        num_combinations=2,
        combination_method='pairwise',
        H1=64,
        H2=256
    )
    
    # Load and prepare data
    filename = 'dataset.xlsx'  # Change this to your file path
    features, labels = tcn.load_and_prepare_data(filename)
    
    # Create feature combinations
    combined_data = tcn.create_feature_combinations()
    
    # Split data
    X_train, X_test, y_train, y_test = train_test_split(
        combined_data, labels, test_size=0.3, random_state=42, stratify=labels
    )
    
    # Train model
    print("Training TCN model...")
    history = tcn.train(X_train, y_train, X_test, y_test, epochs=200)
    
    # Plot training curves
    tcn.plot_training_curves(history)
    
    # Evaluate model
    print("\nEvaluating model...")
    y_pred, y_pred_proba = tcn.evaluate(X_test, y_test)
    
    # Plot confusion matrix
    tcn.plot_confusion_matrix(y_test, y_pred)
    
    # Plot ROC curves
    tcn.plot_roc_curves(y_test, y_pred_proba)
    
    # Feature explanation
    tcn.explain_features(X_test, y_test)
    
    # Save model (optional)
    # tcn.model.save('tcn_model.h5')
    print("\nTraining complete!")


if __name__ == "__main__":

    main()
